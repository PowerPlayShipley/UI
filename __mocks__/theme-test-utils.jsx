// TODO: Fix this??

import * as React from 'react'
import userEvent from '@testing-library/user-event'
import { render as rtlRender } from '@testing-library/react'

import { ThemeProvider } from "styled-components";

import theme from '../src/themes'
import GlobalStyle from "../src/components/GlobalStyle";

function renderWithTheme(ui, { ...options } = {}) {
  const Wrapper = ({ children }) => (
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
      <div id='modal-root' />
    </ThemeProvider>
  )

  return rtlRender(ui, { wrapper: Wrapper, ...options })
}

// THE BELLOW COME FROM `https://github.com/styled-components/styled-components/blob/0aa3170c255a49cd41c3fbeb2b8051b5d132f229/src/test/utils.js`

const stripComments = (str) => str.replace(/\/\*.*?\*\/\n?/g, '');

const stripWhitespace = (str) => str
  .trim()
  .replace(/([;\{\}])/g, '$1  ')
  .replace(/\s+/g, ' ');

const getCSS = (scope) =>
  Array.from(scope.querySelectorAll('style'))
    .map(tag => tag.innerHTML)
    .join('\n')
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

function expectCSSMatches ( _expectation, opts = { ignoreWhitespace: true }) {
  // NOTE: This should normalise both CSS strings to make irrelevant mismatches less likely
  const expectation = _expectation
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

  const css = getCSS(document);

  if (opts.ignoreWhitespace) {
    const stripped = stripWhitespace(stripComments(css));
    expect(stripped).toEqual(stripWhitespace(expectation));
    return stripped;
  } else {
    expect(css).toEqual(expectation);
    return css;
  }
};

export * from '@testing-library/react';

export { renderWithTheme, expectCSSMatches };
