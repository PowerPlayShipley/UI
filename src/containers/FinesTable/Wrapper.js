import styled, { } from "styled-components";

const Wrapper = styled.div`
  display: block;
  max-width: 100%;
  
  overflow: scroll visible;
  
  & table {
    width: 100%;
    height: auto;
    border: none;

    border-collapse: separate;
    border-spacing: 0 8px;
    
    th, td {
      width: 1%;
    }

    th {
      border: none;
      border-right: 1px solid ${({ theme }) => theme.main.colors.border.subtle} !important;
    }

    th:last-child, td:last-child {
      border-right: none !important;
    }
    
    td {
      border: none;
    }
    
    thead {
      font-weight: ${({ theme }) => theme.main.typography.fontWeights.semiBold};
    }
    
    thead {
      th:first-child {
        text-align: left;
      }
    }
    
    tbody tr {
      //--scale: 1.01;
      
      background-color: ${({ theme }) => theme.main.colors.system.white};
      box-shadow: ${({ theme }) => theme.main.misc.shadow.medium }, ${({ theme }) => theme.main.misc.inset.default };
      
      //&:hover {
      //  transform: scale(var(--scale));
      //  
      //  & > * {
      //    transform: scale(calc(1 / var(--scale)));
      //  }
      //}
    }
    
    tbody {
      --border-radius-cell: ${({ theme }) => theme.main.misc.radius["2"]};
      
      tr td:first-child { 
        border-top-left-radius: var(--border-radius-cell);
        border-bottom-left-radius: var(--border-radius-cell);
      }
      tr td:last-child {
        border-top-right-radius: var(--border-radius-cell);
        border-bottom-right-radius: var(--border-radius-cell);
      }
    }
    
    // For only large devices
    tbody td {
      .toolbar {
        visibility: hidden;
        opacity: 0;
        transition: visibility 0s, opacity 0.1s linear;
      }
      
      &:hover {
        .toolbar {
          visibility: visible;
          opacity: 1;
        }
      }
    }
  }
`

export default Wrapper
