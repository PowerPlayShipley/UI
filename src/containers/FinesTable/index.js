import React, { useMemo } from 'react'
import PropTypes from 'prop-types'
import clsx from "clsx";

import Wrapper from "./Wrapper";

import Table from "../../components/Table";
import EditableCell from "../../components/EditableCell";

/**
 * @typedef {Object} RawData
 *
 * @property {string} player
 * @property {[[number]]} games
 * */

/**
 * @param {RawData[]} data
 * @returns {Object}
 * */
const createHeaders = (data) => {
  let games = 1
  data.forEach((previousValue) => {
    games = Math.max(games, previousValue.games.length)
  })

  let remainingHeaders = new Array(games).fill({
    Cell: EditableCell
  }).map((el, idx) => ({ ...el, Header: `Game ${idx + 1}`, accessor: `game${idx + 1}` }))

  return [{
    Header: 'Player',
    accessor: 'player'
  }, ...remainingHeaders]
}

/**
 * @param {RawData[]} data
 * @returns {Object}
 * */
const createData = (data) => {
  return data.map((el) => ({
    player: el.player,
    ...el.games.reduce(((previousValue, currentValue, currentIndex) => {
      return { ...previousValue, [`game${currentIndex+1}`]: currentValue }
    }), { })
  }))
}

const  FinesTable = ({ data: values, onToolbarClick, onItemClick, ...rest }) => {
  const columns = useMemo(() => createHeaders(values), [values])
  const data = useMemo(() => createData(values), [values])

  return (
    <Wrapper {...rest}>
      <Table data={data} columns={columns} onToolbarClick={onToolbarClick} onItemClick={onItemClick} toolbarFloat='right' />
    </Wrapper>
  )
}

export default FinesTable
