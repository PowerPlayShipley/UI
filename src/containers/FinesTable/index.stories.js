import React, { useMemo } from 'react'
import { action } from "@storybook/addon-actions";

import FinesTable from "./index";

export default {
  title: 'Fines/Table',
  component: FinesTable
}

const Template = ({ data, ...args }) => {
  return (
    <div style={{ backgroundColor: '#f5f5f5', padding: '16px 16px' }}>
      <FinesTable data={data} {...args} />
    </div>
  )
};

export const Default = Template.bind({});
Default.args = {
  data: [
    {
      player: 'Ade',
      games: [["H"], ["D", "H", "D", "D"], ["R"]]
    },
    {
      player: 'Tom Collins',
      games: [[], ["H"], ["R"]]
    },
    {
      player: 'Jamie',
      games: [["H", "D", "H", "D", "D"], ["D", "H"], ["R"]]
    }
  ],
  onToolbarClick: action('toolbarClicked'),
  onItemClick: action('itemClick'),
}
