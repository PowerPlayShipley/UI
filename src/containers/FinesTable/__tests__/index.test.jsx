/**
 * @jest-environment jsdom
 */

import React from 'react'
import { renderWithTheme, fireEvent } from '../../../../__mocks__/theme-test-utils'

import FinesTable from "../index";

describe('Display FinesTable', function () {
  it('should create a valid snapshot', function () {
    const data = [
      {
        player: 'Ade',
        games: [["H"], ["D", "H", "D", "D"], ["R"]]
      },
      {
        player: 'Tom Collins',
        games: [[], ["H"], ["R"]]
      },
      {
        player: 'Jamie',
        games: [["H", "D", "H", "D", "D"], ["D", "H"], ["R"]]
      }
    ]

    const { container } = renderWithTheme(<FinesTable data={data} />)
    expect(container.firstChild).toMatchSnapshot()
  });

  it('should create valid buttons', function () {
    const handleClick = jest.fn()

    const data = [
      {
        player: 'Ade',
        games: [
          ['H']
        ]
      }
    ]

    const { getByText } = renderWithTheme(<FinesTable data={data} onItemClick={handleClick} />)

    fireEvent.click(getByText('H').parentNode)
    expect(handleClick).toHaveBeenCalledTimes(1)
  });

  it('should create valid toolbar', function () {
    const handleClick = jest.fn()

    const data = [
      {
        player: 'Ade',
        games: [
          ['H']
        ]
      }
    ]

    const { container } = renderWithTheme(<FinesTable data={data} onToolbarClick={handleClick} />)

    fireEvent.click(container.querySelector('svg'))
    expect(handleClick).toHaveBeenCalledTimes(1)
  });
});
