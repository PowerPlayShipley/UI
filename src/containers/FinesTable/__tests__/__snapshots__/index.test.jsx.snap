// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display FinesTable should create a valid snapshot 1`] = `
.c0 {
  display: block;
  max-width: 100%;
  overflow: scroll visible;
}

.c0 table {
  width: 100%;
  height: auto;
  border: none;
  border-collapse: separate;
  border-spacing: 0 8px;
}

.c0 table th,
.c0 table td {
  width: 1%;
}

.c0 table th {
  border: none;
  border-right: 1px solid rgba(208,215,222,0.5) !important;
}

.c0 table th:last-child,
.c0 table td:last-child {
  border-right: none !important;
}

.c0 table td {
  border: none;
}

.c0 table thead {
  font-weight: 500;
}

.c0 table thead th:first-child {
  text-align: left;
}

.c0 table tbody tr {
  background-color: #ffffff;
  box-shadow: 0 3px 6px rgba(140,149,159,0.15),inset 0 1px 0 rgba(255,255,255,0.25);
}

.c0 table tbody {
  --border-radius-cell: 6px;
}

.c0 table tbody tr td:first-child {
  border-top-left-radius: var(--border-radius-cell);
  border-bottom-left-radius: var(--border-radius-cell);
}

.c0 table tbody tr td:last-child {
  border-top-right-radius: var(--border-radius-cell);
  border-bottom-right-radius: var(--border-radius-cell);
}

.c0 table tbody td .toolbar {
  visibility: hidden;
  opacity: 0;
  -webkit-transition: visibility 0s,opacity 0.1s linear;
  transition: visibility 0s,opacity 0.1s linear;
}

.c0 table tbody td:hover .toolbar {
  visibility: visible;
  opacity: 1;
}

.c1 {
  border-spacing: 0;
  border: 1px solid #d0d7de;
}

.c1 tr:last-child td {
  border-bottom: 0;
}

.c1 th,
.c1 td {
  margin: 0;
  padding: 0.5rem;
  border-bottom: 1px solid #d0d7de;
  border-right: 1px solid #d0d7de;
}

.c1 th:last-child,
.c1 td:last-child {
  border-right: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  padding: 0 0;
}

.c2 .editable {
  overflow: auto;
}

.c2 .toolbar {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  box-shadow: 0 1px 0 rgba(27,31,36,0.04);
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 div[disabled],
.c3 div.disabled,
.c3 div[aria-disabled=true] {
  pointer-events: none;
  color: #8c959f;
  opacity: 65%;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 24px;
  min-width: 24px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 3px;
  padding: 4px;
  color: #222f3e;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c5:hover {
  background-color: #dee0e2;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  max-height: 200px;
  overflow-x: hidden;
  overflow-y: auto;
  padding: 0;
}

<div
  class="c0"
>
  <table
    class="c1"
    role="table"
  >
    <thead>
      <tr
        role="row"
      >
        <th
          colspan="1"
          role="columnheader"
        >
          Player
        </th>
        <th
          colspan="1"
          role="columnheader"
        >
          Game 1
        </th>
        <th
          colspan="1"
          role="columnheader"
        >
          Game 2
        </th>
        <th
          colspan="1"
          role="columnheader"
        >
          Game 3
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Ade
        </td>
        <td
          role="cell"
        >
          <div
            class="c2"
            data="[object Object],[object Object],[object Object]"
            headers="[object Object],[object Object],[object Object],[object Object]"
          >
            <div
              class="c3 editable"
            >
              <div
                class="c4"
              >
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="H"
                >
                  <div
                    class="c6"
                  >
                    H
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5 toolbar"
                  data-class="toolbar"
                  data-item="toolbar"
                >
                  <div
                    class="c6"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-plus fa-w-14 "
                      data-icon="plus"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </td>
        <td
          role="cell"
        >
          <div
            class="c2"
            data="[object Object],[object Object],[object Object]"
            headers="[object Object],[object Object],[object Object],[object Object]"
          >
            <div
              class="c3 editable"
            >
              <div
                class="c4"
              >
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="D"
                >
                  <div
                    class="c6"
                  >
                    D
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="H"
                >
                  <div
                    class="c6"
                  >
                    H
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="D"
                >
                  <div
                    class="c6"
                  >
                    D
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="D"
                >
                  <div
                    class="c6"
                  >
                    D
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5 toolbar"
                  data-class="toolbar"
                  data-item="toolbar"
                >
                  <div
                    class="c6"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-plus fa-w-14 "
                      data-icon="plus"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </td>
        <td
          role="cell"
        >
          <div
            class="c2"
            data="[object Object],[object Object],[object Object]"
            headers="[object Object],[object Object],[object Object],[object Object]"
          >
            <div
              class="c3 editable"
            >
              <div
                class="c4"
              >
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="R"
                >
                  <div
                    class="c6"
                  >
                    R
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5 toolbar"
                  data-class="toolbar"
                  data-item="toolbar"
                >
                  <div
                    class="c6"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-plus fa-w-14 "
                      data-icon="plus"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Tom Collins
        </td>
        <td
          role="cell"
        >
          <div
            class="c2"
            data="[object Object],[object Object],[object Object]"
            headers="[object Object],[object Object],[object Object],[object Object]"
          >
            <div
              class="c3 editable"
            >
              <div
                class="c4"
              >
                <div
                  aria-disabled="false"
                  class="c5 toolbar"
                  data-class="toolbar"
                  data-item="toolbar"
                >
                  <div
                    class="c6"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-plus fa-w-14 "
                      data-icon="plus"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </td>
        <td
          role="cell"
        >
          <div
            class="c2"
            data="[object Object],[object Object],[object Object]"
            headers="[object Object],[object Object],[object Object],[object Object]"
          >
            <div
              class="c3 editable"
            >
              <div
                class="c4"
              >
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="H"
                >
                  <div
                    class="c6"
                  >
                    H
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5 toolbar"
                  data-class="toolbar"
                  data-item="toolbar"
                >
                  <div
                    class="c6"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-plus fa-w-14 "
                      data-icon="plus"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </td>
        <td
          role="cell"
        >
          <div
            class="c2"
            data="[object Object],[object Object],[object Object]"
            headers="[object Object],[object Object],[object Object],[object Object]"
          >
            <div
              class="c3 editable"
            >
              <div
                class="c4"
              >
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="R"
                >
                  <div
                    class="c6"
                  >
                    R
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5 toolbar"
                  data-class="toolbar"
                  data-item="toolbar"
                >
                  <div
                    class="c6"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-plus fa-w-14 "
                      data-icon="plus"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Jamie
        </td>
        <td
          role="cell"
        >
          <div
            class="c2"
            data="[object Object],[object Object],[object Object]"
            headers="[object Object],[object Object],[object Object],[object Object]"
          >
            <div
              class="c3 editable"
            >
              <div
                class="c4"
              >
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="H"
                >
                  <div
                    class="c6"
                  >
                    H
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="D"
                >
                  <div
                    class="c6"
                  >
                    D
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="H"
                >
                  <div
                    class="c6"
                  >
                    H
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="D"
                >
                  <div
                    class="c6"
                  >
                    D
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="D"
                >
                  <div
                    class="c6"
                  >
                    D
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5 toolbar"
                  data-class="toolbar"
                  data-item="toolbar"
                >
                  <div
                    class="c6"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-plus fa-w-14 "
                      data-icon="plus"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </td>
        <td
          role="cell"
        >
          <div
            class="c2"
            data="[object Object],[object Object],[object Object]"
            headers="[object Object],[object Object],[object Object],[object Object]"
          >
            <div
              class="c3 editable"
            >
              <div
                class="c4"
              >
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="D"
                >
                  <div
                    class="c6"
                  >
                    D
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="H"
                >
                  <div
                    class="c6"
                  >
                    H
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5 toolbar"
                  data-class="toolbar"
                  data-item="toolbar"
                >
                  <div
                    class="c6"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-plus fa-w-14 "
                      data-icon="plus"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </td>
        <td
          role="cell"
        >
          <div
            class="c2"
            data="[object Object],[object Object],[object Object]"
            headers="[object Object],[object Object],[object Object],[object Object]"
          >
            <div
              class="c3 editable"
            >
              <div
                class="c4"
              >
                <div
                  aria-disabled="false"
                  class="c5"
                  data-item="R"
                >
                  <div
                    class="c6"
                  >
                    R
                  </div>
                </div>
                <div
                  aria-disabled="false"
                  class="c5 toolbar"
                  data-class="toolbar"
                  data-item="toolbar"
                >
                  <div
                    class="c6"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-plus fa-w-14 "
                      data-icon="plus"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
